File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.js
let CASE_SENSITIVE = false;

function compareStrings(text1, text2){
	let out = false;
	
	if (CASE_SENSITIVE){
		text1 = text1.toUpperCase();
		text2 = text2.toUpperCase();
	}
	
	if (text1 === "" && text2 === ""){
		out = true;
		return out;
	}
	
	if (text1.length != text2.length){
		return out;
	}
	
	for (let i = 0; i < text1.length; i++){
		if (i == 0){
			if (text1[i] == text2[i]){
				out = true;
			} else {
				out = false;
			}
		} else {
			out = out && (text1[i] == text2[i]);
		}
	}
	
	return (out);
}

function firstOccurence (needle, haystack){
	let lenNeedle = needle.length;
	let lenHaystack = haystack.length;
	let out = -1;
	
	if (lenNeedle > lenHaystack){
		out = -1;
	} else if (lenNeedle == lenHaystack) {
		let found = compareStrings(needle, haystack);
		out = found ? 0 : -1;
	} else {
		let lenSearchable = lenHaystack - lenNeedle + 1;
		for (let i = 0; i < lenSearchable; i++){
			if (compareStrings(needle, haystack.slice(i, (i + lenNeedle)))){//complete this 
				out = i;
				break;
			}
		}
	}
	
	return (out);
}

console.log(firstOccurence("ode", "leetcode"));

File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.rb
$CASE_SENSITIVE = false

def compare_strings (text1, text2)
	out = false
	
	if ($CASE_SENSITIVE)
		text1 = text.upcase()
		text2 = text.upcase()
	end
	
	if (text1 == "" and text2 == "")
		out = true
		return out
	end
	
	if (text1.length() != text2.length)
		return out
	end
	
	for i in 0...text1.length()
		if (i == 0)
			if (text1[i] == text2[i])
				out = true
			else
				out = false
			end
		else
			out = out and (text1[i] == text2[i])
		end
	end
	
	return out
end

def first_occurence (needle, haystack)
	len_needle = needle.length()
	len_haystack = haystack.length()
	out = -1
	
	if (len_needle > len_haystack)
		out = -1
	elsif
		found = compare_strings(needle, haystack)
		out = (found) ? 0 : -1
	else
		len_searchable = len_haystack - len_needle + 1
		for i in 0...len_searchable
			if (compare_strings(needle, haystack[i...(i + len_needle)])) #complete this at least
				out = i
				break
			end
		end
	end
	
	return out
end

puts (first_occurence("ode", "leetcode"));
